packages/base-driver/test/unit/wd-proxy/proxy.spec.js

```diff
@@ -0,0 +1,178 @@
+      );
+    });
+    it('should work with urls which do not have session ids', function () {
+      let j = mockProxy({sessionId: '123'});
+      j.getUrlForProxy('http://host.com:1234/session').should.eql(
+        `http://${TEST_HOST}:${port}/session`
+      );
+
+      let newUrl = j.getUrlForProxy('/session');
+      newUrl.should.eql(`http://${TEST_HOST}:${port}/session`);
+    });
+    it('should throw an error if url requires a sessionId but its null', function () {
+      let j = mockProxy();
+      let e;
+      try {
+        j.getUrlForProxy('/session/456/element/200/value');
+      } catch (err) {
+        e = err;
+      }
+      should.exist(e);
+      e.message.should.contain('without session id');
+    });
+    it('should not throw an error if url does not require a session id and its null', function () {
+      let j = mockProxy();
+      let newUrl = j.getUrlForProxy('/status');
+
+      should.exist(newUrl);
+    });
+  });
+
+  describe('straight proxy', function () {
+    it('should successfully proxy straight', async function () {
+      let j = mockProxy();
+      let {status, data} = await j.proxy('/status', 'GET');
+      status.should.equal(200);
+      data.should.eql({value: {foo: 'bar'}});
+    });
+    it('should save session id on session creation', async function () {
+      let j = mockProxy();
+      let {status, data} = await j.proxy('/session', 'POST', {
+        desiredCapabilities: {},
+      });
+      status.should.equal(200);
+      data.should.eql({value: {sessionId: '123', browserName: 'boo'}});
+      j.sessionId.should.equal('123');
+    });
+    it('should overwrite session id when new session generates one', async function () {
+      let j = mockProxy({sessionId: '456'});
+      let {status, data} = await j.proxy('/session', 'POST', {
+        desiredCapabilities: {},
+      });
+      status.should.equal(200);
+      data.should.eql({value: {sessionId: '123', browserName: 'boo'}});
+      j.sessionId.should.equal('123');
+    });
+    it('should pass along request errors', function () {
+      let j = mockProxy({sessionId: '123'});
+      j.proxy('/badurl', 'GET').should.eventually.be.rejectedWith('Could not proxy');
+    });
+    it('should proxy error responses and codes without throwing', async function () {
+      let j = mockProxy({sessionId: '123'});
+      const {status, data} = await j.proxy('/element/bad/text', 'GET');
+      status.should.equal(400);
+      data.should.eql({value: {error: 'invalid element state'}});
+    });
+  });
+
+  describe('command proxy', function () {
+    it('should successfully proxy command', async function () {
+      let j = mockProxy();
+      let res = await j.command('/status', 'GET');
+      res.should.eql({foo: 'bar'});
+    });
+    it('should pass along request errors', function () {
+      let j = mockProxy({sessionId: '123'});
+      j.command('/badurl', 'GET').should.eventually.be.rejectedWith('Could not proxy');
```

Comment:
```suggestion
      j.command('/badurl', 'GET').should.be.rejectedWith('Could not proxy');
```

😆

Reply:
(`eventually` is _only_ needed for assertions which aren't part of `chai-as-promised`.  so you'd use it if, say, you wanted to assert a number _x_ is greater than _y_.)

Reply:
yeah, just copy pasting what was there before, probably written by a slightly dumber me